src/exchanges/populate.ts

```diff
@@ -0,0 +1,331 @@
+  const typeInfo = new TypeInfo(schema);
+  const requiredUserFragments = new Set<FragmentDefinitionNode>();
+  let additionalFragments: Record<string, FragmentDefinitionNode> = {};
+
+  return visit(
+    query,
+    visitWithTypeInfo(typeInfo, {
+      Field: {
+        enter: node => {
+          if (
+            !node.directives ||
+            !node.directives.find(d => d.name.value === 'populate')
+          ) {
+            return;
+          }
+
+          const directives = node.directives.filter(
+            d => d.name.value !== 'populate'
+          );
+
+          const types = getTypes(schema, typeInfo);
+          const newSelections = types.reduce((p, t) => {
+            const typeFrags = typeFragments[t.name];
+            if (!typeFrags) {
+              return p;
+            }
+
+            return [
+              ...p,
+              ...typeFragments[t.name].map(({ fragment }) => {
+                // Add used fragment for insertion at Document node
+                getUsedFragments(fragment).forEach(f =>
+                  requiredUserFragments.add(userFragments[f])
+                );
+
+                // Add fragment for insertion at Document node
+                additionalFragments = {
+                  ...additionalFragments,
+                  [fragment.name.value]: fragment,
+                };
+
+                return {
+                  kind: 'FragmentSpread',
+                  name: {
+                    kind: 'Name',
+                    value: fragment.name.value,
+                  },
+                } as const;
+              }),
+            ];
+          }, [] as FragmentSpreadNode[]);
+
+          const existingSelections =
+            (node.selectionSet && node.selectionSet.selections) || [];
+          const selections =
+            existingSelections.length + newSelections.length !== 0
+              ? [...newSelections, ...existingSelections]
+              : [
+                  {
+                    kind: 'Field',
+                    name: {
+                      kind: 'Name',
+                      value: '__typename',
+                    },
+                  },
+                ];
+
+          return {
+            ...node,
+            directives,
+            selectionSet: {
+              kind: 'SelectionSet',
+              selections,
+            },
+          };
+        },
+      },
+      Document: {
+        leave: node => {
+          return {
+            ...node,
+            definitions: [
+              ...node.definitions,
+              ...Object.values(additionalFragments),
+              ...[...requiredUserFragments],
+            ],
+          };
+        },
+      },
+    })
+  );
+};
+
+/** Get all possible types for node with TypeInfo. */
+const getTypes = (schema: GraphQLSchema, typeInfo: TypeInfo) => {
+  const type = typeInfo.getType();
+
+  if (!type || !('ofType' in type)) {
+    console.warn('PopulateExchange: Unsupported type at populate decorator.');
```

Comment:
I think this is a little cryptic 🤔  can we make this message clearer somehow? I'm not quite sure what it's trying to tell the user about. Also, do we want to log this out identically multiple times?

Reply:
This is something I wanted to talk to you and @imranolas about. I'm assuming we're going to see users adding this decorator on a field with the following return types:
 - GraphQLObjectType
 - GraphQLInterfaceType

Are there any sub/parent/sibling types you can see being used?

Reply:
Oh, I see what it’s used for now then 😅 wasn’t too clear on that.

I’d suspect we’d be able to use it on any field that can have a selection set, so basically: returning objects or unions or interfaces

I feel like that may be it

Reply:
Overlooked unions - adding a test for that now.